draft-ietf-tls-esni.md

```diff
@@ -963,6 +962,135 @@ Variations in the length of the ClientHelloInner ciphertext could leak informati
 about the corresponding plaintext. {{padding}} describes a RECOMMENDED padding
 mechanism for clients aimed at reducing potential information leakage.
 
+## Active Attack Mitigations
+
+This section describes rationale for ECHO properties and mechanics as mitigations
+against active attacks. In all attacks below, the attacker is on-path between
+the target client and server. The goal of the attacker is to learn private information
+in the inner ClientHello, such as the true SNI value.
+
+### Client Reaction Attack Mitigation {#flow-client-reaction}
+
+This attack relies on how clients react to SNI and certificate mismatch as an oracle.
+In particular, the attacker intercepts a legitimate ClientHello and replies with a
+ServerHello, Certificate, CertificateVerify, and Finished messages, wherein the Certificate
+message contains a "test" certificate for the domain name it wishes to query. If the client
+decrypted the Certificate and failed verification (or leaked information about its verification
+process by a timing side channel), the attacker learns that its test certificate name
+was incorrect. As an example, suppose the client's SNI value in its inner ClientHello is
+"example.com," and the attacker replied with a Certificate for "test.com". If the client
+produces a verification failure alert because of the mismatch faster than it would due to
+the Certificate signature validation, information about the name leaks.
+
+~~~
+ Client                         Attacker                      Server
+   ClientHello
+   + key_share
+   + echo        ------>      (intercept)     -----> X (drop)
+
+                             ServerHello
+                             + key_share
+                   {EncryptedExtensions}
+                   {CertificateRequest*}
+                          {Certificate*}
+                    {CertificateVerify*}
+                 <------
+   Alert
+                 ------>
+~~~
+{: #flow-diagram-client-reaction title="Client reaction attack"}
+
+The "echo_nonce" extension in the inner ClientHello prevents this attack. In particular,
+since the attacker does not have access to this value, it cannot produce the right transcript
+and handshake keys needed for encrypting the Certificate message. Thus, the client will fail
+to decrypt before any verification check occurs.
+
+### HelloRetryRequest Hijack Mitigation {#flow-hrr-hijack}
+
+This attack aims to exploit server HRR state management to recover information about
+a legitimate ClientHello using its own adversarially-controlled ClientHello.
+To begin, the attacker intercepts and forwards a legitimate ClientHello with an
+"encrypted_client_hello" (echo) extension to the server, which triggers a HelloRetryRequest
+in return. Rather than forward the retry to the client, the attacker, attempts to
```

Comment:
I think you need to say something about how this had to legitimately be an HRR case. Or can the attacker force it?

Reply:
Good point. I'll call out that assumption explicitly.